dabuild.in

```diff
@@ -58,10 +58,13 @@ else
   esac
 fi
 
-## check $DABUILD_PACKAGES is a writable directory
 ABUILD_PACKAGES=${DABUILD_PACKAGES:-${PWD%/aports/*}/packages/${ABUILD_VERSION}}
 
-mkdir -p $ABUILD_PACKAGES
+# docker will create mount points as root if not available so create them first
+# use uid 1000 as this is the uid of the builder user in the container
+install -d -o 1000 -g 1000 "$ABUILD_PACKAGES" "$HOME"/.abuild
```

Comment:
Whoops -- thanks for catching that I don't check `"$HOME/".abuild` exists.

Not sure about creating them with [UG]ID=1000 though. Maybe better to invoke `docker run ...` below with `--user "$(id -u)":"$(id -g)"`? (I'm only on D4M here and it looks like some magic is being done to map through to my host user account in either case, so can't easily test.)

Reply:
Why is setting explicitly 1000 a problem? the first user is always assigned 1000?

Reply:
hmm, i was thinking only from a point of view from the container, its probably not smart to set 1000 if the current host user has a different uid (which is possible). I guess we need to rethink how to manage those bind mounts or instead use docker volumes.

Reply:
Yes-- mapping between the uid spaces in container and on host is always a fiddle. 

AIUI, in general there are uids and gids defined in the container, some (typically, most) with corresponding user and group names; and there are uids and gids similarly on the host. None of these need to overlap. When the image is built, I create the `builder(1000):builder(1000)` user and that will be the user by which commands are executed by default.

At this point, the script is running on the host so should (IMO) be creating files/directories using the default host's uid:gid (ie., no `-o` or `-g` switches to `install`).

When we `docker run` then ideally we would create files using the `uid:gid` of the user on the host. But the CLI mechanism for doing that (AIUI) is to use the numeric IDs by passing in `--user $(id -u):$(id -g)`. Unfortunately, that may prevent use of `sudo` inside the container because there is a good chance that the uid that ends up being used will not exist in `/etc/passwd` in the container and so `sudo` will refuse to run.

Perhaps a compromise would be to leave the uid in the container as-is (ie., `builder(1000)`) but to add something like `--group-add $(id -G | sed 's/ / --group-add /g')"` to the `docker run ...` command line -- assuming directories are created with `g+w` then at least file creation from inside the container will work while still letting us use `sudo` inside the container...?

(Unfortunately I can't test any of this as D4M does magic to make the mappings work behind my back -- on Mac/Windows it's extra complicated because there are effectively three layers of namespace involved: the actual host (Mac/Windows), the Linux VM, container.)

Reply:
(Or we could just give up using bind mounts at all, and use named volumes for everything. But extracting built packages becomes trickier then. Maybe that doesn't matter though?)